jmeter-service/testrunner.go

```
@@ -0,0 +1,269 @@
+	return &TestRunner{eventSender}
+}
+
+// RunTests downloads the JMeter configuration, eventually run a basic health check
+// and executes the Jmeter test
+func (tr *TestRunner) RunTests(testInfo TestInfo) error {
+	if err := tr.sendTestsStartedEvent(testInfo); err != nil {
+		logger.WithError(err).Error("Unable to send test '.started' event")
+		return err
+	}
+	testStartedAt := time.Now()
+
+	jmeterConfig, err := getJMeterConf(testInfo)
+	if err != nil {
+		return err
+	}
+
+	if err := tr.runHealthCheck(testInfo, testStartedAt, jmeterConfig); err != nil {
+		return err
+	}
+
+	res, err := tr.runTests(testInfo, jmeterConfig)
+	if err != nil {
+		if err := tr.sendErroredTestsFinishedEvent(testInfo, testStartedAt, err.Error()); err != nil {
+			logger.WithError(err).Errorf(errMsgSendFinishedEvent, testInfo.ToString())
+		}
+	}
+	msg := fmt.Sprintf("Tests for %s with status = %s.%s", testInfo.TestStrategy, strconv.FormatBool(res), testInfo.ToString())
+	if !res {
+		if err := tr.sendTestsFinishedEvent(testInfo, testStartedAt, msg, keptnv2.ResultFailed); err != nil {
+			logger.WithError(err).Errorf(errMsgSendFinishedEvent, testInfo.ToString())
+		}
+	} else {
+		if err := tr.sendTestsFinishedEvent(testInfo, testStartedAt, msg, keptnv2.ResultPass); err != nil {
+			logger.WithError(err).Errorf(errMsgSendFinishedEvent, testInfo.ToString())
+		}
+	}
+	return nil
+}
+
+func (tr *TestRunner) runTests(testInfo TestInfo, jmeterConf *JMeterConf) (bool, error) {
+	var testStrategy = strings.ToLower(testInfo.TestStrategy)
+
+	if testStrategy == "" {
+		logger.Info("No testStrategy specified therefore skipping further test execution and sending back success")
+		return true, nil
+	}
+
+	teststrategyWorkload, err := getWorkloadForStrategy(jmeterConf, testStrategy)
+	if err != nil {
+		return false, err
+	}
+	if teststrategyWorkload == nil {
+		logger.Errorf("No workload definition found for testStrategy %s", testStrategy)
+		return false, nil
```

should we also return an error here? Because this case is actually covered by the `getWorkloadForStrategy()` function as it returns `fmt.Errorf("no workload configuration found for teststrategy: %s", teststrategy)` if no workload is found. Double checking wether the `testStrategyWorkload` is really `!= nil` is still a good idea though